fix: add a blog on using the tekton plugin (#4269)
jenkins-infra/jenkins.io: content/blog/2021/04/2021-04-20-tekton-plugin.adoc

---
layout: post
title: "Easily reuse Tekton and Jenkins X from Jenkins"
tags:
- jenkins-x
- kubernetes
- pipeline
- tekton

> - gitops

Reviewer: ```suggestion
- gitops
- interoperability
```

Might be used by @MarckK for such kind of blogposts and quick search in interoperability SIG

authors:
- jstrachan
description: >
  This post gives an overview of the new Tekton plugin for Jenkins and how it can help you reuse the power of Tekton pipelines, the Tekton Catalog and the automated CI/CD from Jenkins X

> ---

Reviewer: As discussed with @jstrachan , will add this one:

![image](https://user-images.githubusercontent.com/3000480/115572801-87c22400-a2c0-11eb-9d5c-af4c410fa735.png)

![image](https://user-images.githubusercontent.com/3000480/115572756-7e38bc00-a2c0-11eb-8308-e03fcf4aeb20.png)

== What is Tekton?

> image:/images/tekton-pipline/logo.png[TektonLogo,width=200,role="left"]

Reviewer 2: Spelling fix
```suggestion
image:/images/tekton-plugin/logo.png[TektonLogo,width=200,role="left"]
```

link:https://tekton.dev/[Tekton] is a powerful and flexible open-source framework for creating CI/CD systems, allowing developers to build, test, and deploy across cloud providers and on-premise systems.

=== Why use Tekton?

Tekton pipelines have a number of benefits:

* they are cloud native and designed from the ground up for kubernetes
* each Tekton `Pipeline` is fully declarative and completely self described; it does not depend on any separate out of band Jenkins controllers, plugins or plugin/controller configurations
* each `Pipeline Task` runs as a stand alone kubernetes link:https://kubernetes.io/docs/concepts/workloads/pods/[Pod] which is completely independent of any other pods and pipelines and are fully scheduled by Kubernetes to maximise resilience and optimize resource usage. A bad pipeline cannot take down another one & the kubernetes scheduler manages them all
* each step can be any command in any container image with whatever secrets, volume mounts, environment variables and resource limits you need
* there is no need to bundle a JVM or Jenkins Remoting container into the pod so you can keep resources and cost down

== Why use Jenkins and Tekton together?

> link:https://www.jenkins.io/[Jenkins] is the most popular open source automation server around. Lots of developers use it every day to get things done.

Reviewer: ```suggestion
Jenkins is the most popular open source automation server around. Lots of developers use it every day to get things done.
```

> 

Reviewer: ```suggestion
```

Jenkins can now be used to automate link:https://tekton.dev/[Tekton] pipelines too which helps teams digitally transform to more cloud native solutions for their CI and CD.

> 

Reviewer: ```suggestion
In such a case, you can use Tekton pipeline engine while getting all benefits from Jenkins as an orchestrator, user interface and the reporting engine.
```

Or just a "control plane"?

== Introducing the Tekton Plugin for Jenkins

> The link:https://github.com/jenkinsci/tekton-client-plugin[tekton plugin for jenkins] lets you easily use Jenkins to automate creating and running Tekton pipelines.

Reviewer: Add a few more words from the readme or a brief from @waveywaves ?

```suggestion
The plugin:tekton-client[Tekton Client plugin for Jenkins] lets you easily use Jenkins to automate creating and running Tekton pipelines.
It bridges the Kubernetes learning gap and allows invoking Tekton Pipelines and resources through Jenkins.
This allows users to not have much of the Kubernetes specific knowledge beforehand and work.
```

> Its a single Jenkins plugin to install - so it's easy to use.

Reviewer: Would it makes sense to reference https://cd.foundation/blog/2020/11/05/bridging-the-gap-with-tekton-client-plugin-for-jenkins/ from @waveywaves or to embed the video?

Author: I've added a link to the blog post and to @waveywaves

We have found on the link:https://jenkins-x.io/[Jenkins X project] that link:https://jenkins-x.io/blog/2021/02/25/gitops-pipelines/[a nice way to do this via GitOps] such that we reference versioned Tekton Tasks and Pipelines in git so that they are easy to link:https://jenkins-x.io/v3/develop/pipelines/catalog/#referencing-a-task-or-step[reuse] or link:https://jenkins-x.io/v3/develop/pipelines/catalog/#overriding-a-pipeline-step-locally[override].

So we reuse Tasks and Pipelines via the link:https://jenkins-x.io/v3/develop/pipelines/catalog/#sourceuri-notation[uses: image notation] which lets us keep all of our Tekton Tasks and Pipelines in vanilla Tekton YAML; so that the IDE completion and validation works - but we can easily reuse Tasks or steps from libraries while also link:https://jenkins-x.io/v3/devops/patterns/version_everything/[Versioning Everything]

**Note** that if wish to reuse steps/tasks via the link:https://jenkins-x.io/v3/develop/pipelines/catalog/#sourceuri-notation[uses: image notation] then you must click the `Tekton Catalog` flag in your Job definition which will then resolve the `uses:` clause with the actual step/task.

> == What is Jenkins X?

Reviewer: It might make sense to split to two blogpost, but no strong opinion

Author: I'd rather it just be 1 really as then I don't have to create another demo ;)

> == Whats is Jenkins X

Reviewer 2: Typo?
```suggestion
== What is Jenkins X?
```

* link:https://jenkins-x.io/v3/develop/environments/preview/[Preview Environments] lets you propose code changes via Pull Requests and have a Preview Environment automatically created, running your code in kubernetes to get fast feedback from your team before agreeing to merge changes to the main branch
* link:https://jenkins-x.io/v3/develop/developing/#using-chatops[ChatOps] comment on Pull Requests to give feedback, approve/hold changes, trigger optional pipelines for additional testing and other link:https://jenkins-x.io/v3/develop/reference/chatops/[ChatOps commands]

All of the above is implemented in reusable Tekton pipelines.

=== Reusing Jenkins X Pipelines

So how can we reuse automated CI/CD pipelines from link:https://jenkins-x.io/[Jenkins X project] from Jenkins?

> Make sure you have the link:https://github.com/jenkinsci/tekton-client-plugin[tekton plugin for jenkins] installed in your Jenkins server.

Reviewer: ```suggestion
Make sure you have the plugin:tekton-client[Tekton Client plugin for Jenkins] installed in your Jenkins server.
```

==== Using a working template

If you want to start with a working example then

* link:https://github.com/jstrachan/node-example/generate[Create A Git Repository From This Template]

* add a new `Frestyle project` to your Jenkins server
* enable the `Git` source code management for your new github.com repository
* click `Add build Step` (near the bottom of the page) and then select `Tekton : Create Resource (Raw)`
* make sure that `FILE` is selected for the input and enter the name `.lighthouse/jenkins-x/release.yaml` for the file name
* if you are using a Jenkins X cluster enter `jx` for the namespace
* ensure that `Enable Tekton Catalog` is checked
* now save the pipeline - it should look something like this:

> image:/images/tekton-plugin/example.png["Jenkins Console"]

Reviewer: Likely the image could be also compressed

Author: I've shrunk it 50% for now - that ok?

Reviewer: Yes, thanks! Soon we should have a stable GitHub action for that

> image:/images/tekton-plugin/example.png["Jenkins Console"]

Author: any idea how to do it? :) I know what the generated XML looks like; but never made a Freestyle job before via CasC

* you have to maintain 2 files: the `Jenkinsfile` and the `pod.yaml` and it's hard to share/override both of those files across multiple repositories as you need to make changes (e.g. overriding environment variables/images/commands/resource limits on demand on steps).

Though one downside of the tekton approach is that by default there is no automatic synchronisation of state; after a Task in tekton completes there's no automatic upload of state to the Jenkins controllers disk. You can always add a step in your Task to upload workspace state to the Jenkins controller if that's what you want.

Though remember that tekton plugin doesn't take anything away; so you can mix and match the kubernetes and tekton plugins to suit your needs.

== Conclusion

We are really excited about the combination of Jenkins, link:https://tekton.dev/[Tekton] and link:https://jenkins-x.io/[Jenkins X] letting developers pick the best tool for the job while becoming more cloud native and increasing the automation help reduce the amount of manual work creating and maintaining pipelines while also helping to improve the quality and practices of our CI/CD.

> Please try it out and link:https://github.com/jenkinsci/tekton-client-plugin/issues[let us know how you get on]

Reviewer: Add a link to https://github.com/jenkinsci/tekton-client-plugin/blob/master/CONTRIBUTING.md as well?

---
layout: post

> title: "Easily reuse Tekton and Jenkins X from Jenkins"

Reviewer: or "Easily reuse Tekton from Jenkins and Jenkins X"? It does not look like we reuse Jenkins X from Jenkins, but I might be wrong

Author: the title is correct, we can reuse Tekton and/or Jenkins X (pipelines) from Jenkins
